mcstatus/motd.py

```diff
@@ -0,0 +1,558 @@
+            elif isinstance(element, TranslationTag):
+                continue
+            elif isinstance(element, WebColor):
+                result += f"<span style='color:{element.hex}'>"
+                on_reset += "</span>"
+
+            elif isinstance(element, Formatting):
+                if element == Formatting.RESET:
+                    result += on_reset
+                    on_reset = ""
+                elif element == Formatting.OBFUSCATED:
+                    result += "<span class=obfuscated>"
+                    on_reset += "</span>"
+                else:
+                    tag_name = FORMATTING_TO_HTML_TAGS[element]
+                    result += "<" + tag_name + ">"
+                    on_reset += "</" + tag_name + ">"
+            else:
+                mc_color_to_hex = MINECRAFT_COLOR_TO_HEX_BEDROCK if self.bedrock else MINECRAFT_COLOR_TO_HEX_JAVA
+                foreground_color, background_color = mc_color_to_hex[element]
+
+                result += f"<span style='color:{foreground_color};text-shadow:0 0 1px {background_color}'>"
+                on_reset += "</span>"
+
+        return result + on_reset + "</p>"
+
+    @property
+    def xterm_256(self) -> str:
+        """Getter for a xterm-256 variant of MOTD.
+
+        The only difference from ``ansi`` is that it's using RGB colors instead of ``red``, ``blue`` etc.
+
+        :returns: String full of xterm-256 colors.
+        """
+        result = r"\e[0m"
+        for element in self.parsed:
+            if isinstance(element, str):
+                result += element
+            elif isinstance(element, TranslationTag):
+                continue
+            elif isinstance(element, WebColor):
+                result += r"\x1b[38;2;{0};{1};{2}m".format(*element.rgb)
+
+            elif isinstance(element, Formatting):
+                if element == Formatting.RESET:
+                    result += r"\e[0m"
+                else:
+                    tag_number = FORMATTING_TO_ANSI_TAGS[element]
+                    result += r"\e[" + tag_number + "m"
+            else:
+                mc_color_to_hex = MINECRAFT_COLOR_TO_HEX_BEDROCK if self.bedrock else MINECRAFT_COLOR_TO_HEX_JAVA
+                foreground_color, background_color = mc_color_to_hex[element]
+
+                result += r"\x1b[38;2;{0};{1};{2}m".format(
+                    *tuple(int(foreground_color.lstrip("#")[i : i + 2], 16) for i in (0, 2, 4))
+                )
+                result += r"\x1b[48;2;{0};{1};{2}m".format(
+                    *tuple(int(background_color.lstrip("#")[i : i + 2], 16) for i in (0, 2, 4))
+                )
+
+        return result + r"\e[0m"
+
+    @property
+    def ansi(self) -> str:
+        """Getter for an ANSI colors variant of MOTD.
+
+        The only difference from ``xterm_256`` is that it's using ``red``, ``blue`` etc. instead of RGB colors.
+        This also using ``xterm_256`` colors, if found ``WebColor`` in ``Motd.parsed``.
```

Comment:
```suggestion

        .. note:: If ``Motd.parsed`` contains ``WebColor`` items. as they're purely RGB based, this will still use ``xterm_256`` colors with these.
```

However, can't we just approximate this down? Full RGB colors can always be scaled down, yeah we'll loose color quality, but that's kind of the point with ANSI, otherwise we'd just use xterm_256.

Reply:
We can't. Because there are only 16 ANSI colors and 16,7 millions RGB colors, so down scaling will work terrible. I tried it:
![image](https://user-images.githubusercontent.com/68118654/206923971-edf3bb90-843a-4e41-acd8-4f1d956017f2.png)


We also can't use `ANSI 256-bit` because it is also known as `xterm-256`.

Reply:
Hmm, I think there has been some confusion between xterm-256 and ansi.

- ANSI escape sequences are a standard that expresses how colors should be represented using some special escape sequences.
- xterm-256 is just using ANSI escape sequences, expressing one of 256 colors (meaning these are 8-bit colors). This color range is often supported by almost all terminals.
- However, some terminals, especially in the early ages, only gave us support for 16 colors (4-bit color)
- Though, with modern terminals, they actually have truecolor or 24-bit color support, meaning you can use ANSI sequences that represent full standard RGB spectrum (being 3 bytes, 8-bits per each color (red, green, blue)).

For more details, see: <https://en.wikipedia.org/wiki/ANSI_escape_code>

Because of this, it the naming here doesn't really make sense. I think kevin mentioned xterm_256 color support, and I originally didn't really know what that was either so I just went with it and wrote it off as simply another convention. However as it's just a specific form of ANSI sequences, with the term itself being used to express this functionality on a terminal, it's really weird to have a `xterm_256` property here.

Instead, what we could do is either:
- Make `ansi` a function, taking in an argument specifying either 24-bit color (truecolor, full RGB spectrum, what is currently being done in `xterm_256`), 8-bit color (256 colors in total, what xterm_256 should've been, what's currently being done with `ansi` property, though without the fallback to 256), and perhaps also 4-bit color (only 16 colors in total, some TTYs still use this)
- Make a bunch of properties: `ansi_24`, `ansi_8` and `ansi_4`

As to the downscaling, you've likely just miss-implemented the downscaling here, check <https://gist.github.com/MicahElliott/719710> for a reference implementation in python (note that it's python 2 though, but it's mostly compatible).

Obviously, the loss of color quality will still be huge, but the colors do at least look similar. You certainly won't get a green color from blue. Check <https://www.ditig.com/256-colors-cheat-sheet> to see all of the 256-color numbers that you can use (note that the first 0-15 colors are actually the 4-bit colors. Here's a quick reference if you need one:
![image](https://user-images.githubusercontent.com/20902250/206933802-5c53506f-34c1-4647-b5a0-3d5c8d2cea8b.png)

Reply:
Thanks for the explanation, I will test the new algorithm later.

I would also prefer property names `ansi_24bit`, `ansi_8bit` and `ansi_4bit`.

Reply:
Well, it doesn't work well
```
$ python2 colortrans.py 255
xterm color 255 -> RGB exact eeeeee
$ python2 colortrans.py eeeeee
RGB eeeeee -> xterm color approx 231 (ffffff)
```
But this can be fixed by one if check.

Reply:
Why python 2? I thought this project only supported python 3

Reply:
That's just the reference script I found that is able to downscale RGB colors to 256 colors. It was written in python 2. Obviously, I'm not suggesting to uses the python 2 version here, it will need to be adapted to fit our code-base, I just mentioned it for reference on how this should be done.  @CoolCat467

Reply:
Resolve it please, as we no longer support ANSI <24 bits.